### ISSUE TITLE
Can't build docker image on MacOS Sierra

### ISSUE THREAD

**Issue by gsc-dev** [reactioncommerce/reaction#1946]
For bugs, please fill out the three following sections completely. Failure to provide
this information may result in your ticket being closed without comments.

Are you looking for support for installation/getting started with Reaction? Please go
to our [Gitter](https://gitter.im/reactioncommerce/reaction) channel instead.

### Expected behavior
A successful docker build

### Actual Behavior

```
[-] Building Meteor application...

/root/.meteor/packages/meteor-tool/.1.4.2_7.z6ybt6++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/fibers/future.js:280
						throw(ex);
						^

Error: EXDEV: cross-device link not permitted, rename '/opt/reaction/dist/bundle' -> '/opt/reaction/dist/bundle-garbage-k0lysc'
    at Error (native)
    at Object.fs.renameSync (fs.js:681:18)
    at Object.wrapper [as rename] (/tools/fs/files.js:1535:35)
    at Object.files.renameDirAlmostAtomically (/tools/fs/files.js:965:11)
    at Builder.complete (/tools/isobuild/builder.js:690:13)
    at /tools/isobuild/bundler.js:2519:13
    at /tools/isobuild/bundler.js:2426:7
    at /tools/isobuild/bundler.js:2827:22
    at /tools/utils/buildmessage.js:271:13
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at bundle (/tools/isobuild/bundler.js:2660:31)
    at /tools/isobuild/bundler.js:2607:32
    at Object.withCache (/tools/fs/files.js:1601:12)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2607:16)
    at buildCommand (/tools/cli/commands.js:952:30)
    at Command.func (/tools/cli/commands.js:827:12)
    at /tools/cli/main.js:1483:23

```

### Steps to Reproduce the Behavior

sudo npm i -g reaction-cli
reaction init
cd reaction
reaction
docker build . -t reaction

```
sudo npm i -g reaction-cli
reaction init
cd reaction
reaction

Using settings file at settings/dev.settings.json

Setting up plugin imports...

Setting up style imports...

[[[[[ ~/code/reaction ]]]]]

=> Started proxy.
=> Meteor 1.4.3.1 is available. Update this project with 'meteor update'.
=> Started MongoDB.
cfs:gridfs: updating npm dependencies -- mongodb, gridfs-stream...
15:12:31.159Z  INFO Reaction: Load default data from /private/data/
15:12:31.400Z  INFO Reaction: JobServer started
15:12:31.403Z  WARN Reaction: Skipped loading settings from reaction.json.
15:12:35.368Z  INFO Reaction: Using meteor --settings to create admin user
15:12:37.905Z  WARN Reaction:
  
   *********************************
          
    IMPORTANT! DEFAULT ADMIN INFO
          
    EMAIL/LOGIN: n4ssnyp8@localhost
          
    PASSWORD: UhryOaHj
          
   ********************************* 
  
  
15:12:37.963Z  INFO Reaction: Migrating from version 0 -> 3
15:12:37.963Z  INFO Reaction: Running up() on version 1
15:12:38.073Z  INFO Reaction: Running up() on version 2
15:12:38.113Z  INFO Reaction: Running up() on version 3
15:12:38.142Z  WARN Reaction: Skipped loading settings from reaction.json.
15:12:40.191Z  INFO Reaction: Finished migrating.
15:12:40.282Z  WARN Reaction: OpenExchangeRates API not configured. Not adding fetchRates job
15:12:40.284Z  WARN Reaction: OpenExchangeRates API not configured. Not adding flushRates job
15:12:40.510Z  INFO Reaction: Reaction initialization finished.
=> Started your app.

=> App running at: http://localhost:3000/
^C


docker build . -t reaction
Sending build context to Docker daemon  7.24 MB
Step 1/3 : FROM reactioncommerce/base:v1.2.0
# Executing 2 build triggers...
Step 1/1 : COPY . $APP_SOURCE_DIR
Step 1/1 : RUN cd $APP_SOURCE_DIR &&             bash $BUILD_SCRIPTS_DIR/build-meteor.sh &&             bash $BUILD_SCRIPTS_DIR/post-build-cleanup.sh
 ---> Running in 23d0e74ced1a
/opt/nodejs/bin/reaction -> /opt/nodejs/lib/node_modules/reaction-cli/dist/main.js
/opt/nodejs/bin/rc -> /opt/nodejs/lib/node_modules/reaction-cli/dist/main.js
reaction-cli@0.7.6 /opt/nodejs/lib/node_modules/reaction-cli
├── semver@5.3.0
├── uuid@3.0.1
├── chalk@1.1.3 (escape-string-regexp@1.0.5, supports-color@2.0.0, ansi-styles@2.2.1, has-ansi@2.0.0, strip-ansi@3.0.1)
├── fs-extra@2.0.0 (graceful-fs@4.1.11, jsonfile@2.4.0)
├── node-fetch@1.6.3 (is-stream@1.1.0, encoding@0.1.12)
├── update-notifier@2.1.0 (is-npm@1.0.0, semver-diff@2.1.0, lazy-req@2.0.0, xdg-basedir@3.0.0, configstore@3.0.0, latest-version@3.0.0, boxen@1.0.0)
├── shelljs@0.7.6 (interpret@1.0.1, glob@7.1.1, rechoir@0.6.2)
├── yargs@7.0.1 (get-caller-file@1.0.2, decamelize@1.2.0, camelcase@3.0.0, y18n@3.2.1, which-module@1.0.0, set-blocking@2.0.0, yargs-parser@5.0.0, require-main-filename@1.0.1, require-directory@2.1.1, cliui@3.2.0, string-width@1.0.2, os-locale@1.4.0, read-pkg-up@1.0.1)
├── analytics-node@2.1.1 (clone@2.1.0, join-component@1.1.0, crypto-token@1.0.1, superagent-retry@0.6.0, component-type@1.2.1, commander@2.9.0, debug@2.6.1, superagent@3.5.0, superagent-proxy@1.0.2)
├── inquirer@3.0.6 (ansi-escapes@1.4.0, through@2.3.8, cli-width@2.1.0, strip-ansi@3.0.1, mute-stream@0.0.7, figures@2.0.0, string-width@2.0.0, run-async@2.3.0, external-editor@2.0.1, cli-cursor@2.1.0, rx@4.1.0)
├── lodash@4.17.4
└── babel-polyfill@6.23.0 (regenerator-runtime@0.10.3, babel-runtime@6.23.0, core-js@2.4.1)

[-] Running Reaction plugin loader...


Setting up plugin imports...

Setting up style imports...

Done!


[-] Running npm install in app directory...

npm WARN deprecated node-uuid@1.4.7: use uuid module instead
npm WARN prefer global marked@0.3.6 should be installed with -g
npm WARN prefer global npm@2.15.11 should be installed with -g

> fibers@1.0.15 install /opt/reaction/src/node_modules/fibers
> node build.js || nodejs build.js

`linux-x64-46` exists; testing
Binary is fine; exiting

> dtrace-provider@0.8.0 install /opt/reaction/src/node_modules/dtrace-provider
> node scripts/install.js


> bcrypt@1.0.2 install /opt/reaction/src/node_modules/bcrypt
> node-pre-gyp install --fallback-to-build

node-pre-gyp ERR! Tried to download(404): https://github.com/kelektiv/node.bcrypt.js/releases/download/v1.0.2/bcrypt_lib-v1.0.2-node-v46-linux-x64.tar.gz 
node-pre-gyp ERR! Pre-built binaries not found for bcrypt@1.0.2 and node@4.7.3 (node-v46 ABI) (falling back to source compile with node-gyp) 
make: Entering directory '/opt/reaction/src/node_modules/bcrypt/build'
  CXX(target) Release/obj.target/bcrypt_lib/src/blowfish.o
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt.o
  CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt_node.o
  SOLINK_MODULE(target) Release/obj.target/bcrypt_lib.node
  COPY Release/bcrypt_lib.node
  COPY /opt/reaction/src/node_modules/bcrypt/lib/binding/bcrypt_lib.node
  TOUCH Release/obj.target/action_after_build.stamp
make: Leaving directory '/opt/reaction/src/node_modules/bcrypt/build'

> spawn-sync@1.0.15 postinstall /opt/reaction/src/node_modules/spawn-sync
> node postinstall


> radium@0.18.1 postinstall /opt/reaction/src/node_modules/radium
> cd lib || npm run lib

reaction@0.19.1 /opt/reaction/src
+-- @reactioncommerce/authorize-net@1.0.8 
| +-- 42-cent-base@1.0.0 
`<snip>`
  |   `-- core-js@1.2.7 
  `-- wgxpath@1.0.0 


[-] Building Meteor application...

/root/.meteor/packages/meteor-tool/.1.4.2_7.z6ybt6++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/fibers/future.js:280
						throw(ex);
						^

Error: EXDEV: cross-device link not permitted, rename '/opt/reaction/dist/bundle' -> '/opt/reaction/dist/bundle-garbage-1jv8jjx'
    at Error (native)
    at Object.fs.renameSync (fs.js:681:18)
    at Object.wrapper [as rename] (/tools/fs/files.js:1535:35)
    at Object.files.renameDirAlmostAtomically (/tools/fs/files.js:965:11)
    at Builder.complete (/tools/isobuild/builder.js:690:13)
    at /tools/isobuild/bundler.js:2519:13
    at /tools/isobuild/bundler.js:2426:7
    at /tools/isobuild/bundler.js:2827:22
    at /tools/utils/buildmessage.js:271:13
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:264:29
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:262:18
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at /tools/utils/buildmessage.js:253:23
    at [object Object].withValue (/tools/utils/fiber-helpers.js:89:14)
    at Object.capture (/tools/utils/buildmessage.js:252:19)
    at bundle (/tools/isobuild/bundler.js:2660:31)
    at /tools/isobuild/bundler.js:2607:32
    at Object.withCache (/tools/fs/files.js:1601:12)
    at Object.exports.bundle (/tools/isobuild/bundler.js:2607:16)
    at buildCommand (/tools/cli/commands.js:952:30)
    at Command.func (/tools/cli/commands.js:827:12)
    at /tools/cli/main.js:1483:23
The command '/bin/sh -c cd $APP_SOURCE_DIR &&             bash $BUILD_SCRIPTS_DIR/build-meteor.sh &&             bash $BUILD_SCRIPTS_DIR/post-build-cleanup.sh' returned a non-zero code: 1
```

For enhancements or other issues, please describe in as much detail as
possible how the app should be changed and how this would benefit
users/developers

**Comment by jshimko** [comment]
Yep, this is a known issue with Meteor meteor/meteor#7852

You can also read more about my current understanding of the cause in one of my personal repos...
https://github.com/jshimko/meteor-launchpad/issues/19#issuecomment-280853956

This issue has been around for a long time when building on CoreOS, but the 1.13 release of Docker started making it happen on macOS too.  Coincidentally, I'm working on this right now.  We're likely going to end up running a fork of Meteor inside our base image.  Assuming I can fix it, hopefully they accept a PR for the update.

Since we surprisingly don't already have a dedicated issue open for this yet, I'll leave this open and follow up with progress here.  Thanks for taking the time to report it.

**Comment by gsc-dev** [comment]
Thanks for letting me know @jshimko 

Good luck with your fork 👍

**Comment by jshimko** [comment]
For anyone following along, I figured out a solution for this last week and I submitted a PR to Meteor.  I have no idea if/when it will get merged, but at least now we know there's an option.  

https://github.com/meteor/meteor/pull/8491

**Comment by agnesv2** [comment]
Just wanted to mention that I got the same error when trying to build on win10 today. Will try the solution.

**Comment by jshimko** [comment]
@agnesv2 We just started seeing some action on the Meteor PR, so they may actually merge my fix soon.  Doing the workaround manually yourself is not a trivial amount of work (especially on Windows) because you have to run a fork of Meteor and [build a new dev bundle](https://github.com/meteor/meteor/blob/devel/Development.md#the-dev-bundle).  Your best bet is probably to keep an eye on that PR and wait for the fix.  

However, if it's an emergency (or you just don't want to wait), you can also use my PR branch in my fork.

```sh
# clone my fix
git clone -b meteor-issue-7852 https://github.com/jshimko/meteor.git

cd meteor

# create the dev bundle
./scripts/generate-dev-bundle.sh

# run meteor from that checkout 
./path/to/forked/meteor build --directory /path/to/bundle/dir
```

It's not pretty, but it definitely fixes the problem.  I'd still just wait though.  ;)

**Comment by agnesv2** [comment]
Ok, thank you very much for explaining how to use your fix! Much appreciated.

**Comment by jshimko** [comment]
My fix was merged over at meteor/meteor#8491 and tagged for the `Release 1.4.4` milestone.

Once Meteor 1.4.4 lands (hopefully soon), this issue should be gone!

**Comment by aaronjudd** [comment]
Looks like that fix has been reverted and they're trying a new approach in https://github.com/meteor/meteor/pull/8569

**Comment by owenhoskins** [comment]
Works for me following the above now that it is running Meteor 1.4.4.1!

**Comment by aaronjudd** [comment]
@jshimko can we resolve this?

**Comment by jshimko** [comment]
Yes, the cross device build error appears to be fixed.

**Comment by aaronjudd** [comment]
I'm still experiencing and seeing issue building on Sierra...

**Comment by jshimko** [comment]
@aaronjudd @prakspcd Are you specifically seeing `Error: EXDEV: cross-device link not permitted, rename`?  

I get an issue with the build (another Meteor file system headache on macOS), but it's not the same issue from this PR.  I haven't been able to make that happen since the fix Meteor released a couple months back.

**Comment by gsc-dev** [comment]
Hi @jshimko 

I am also still having issues during the building phase...

```[-] Building Meteor application...


Even with METEOR_ALLOW_SUPERUSER or --allow-superuser, permissions in your app
directory will be incorrect if you ever attempt to perform any Meteor tasks as
a normal user. If you need to fix your permissions, run the following command
from the root of your project:

  sudo chown -Rh <username> .meteor/local

cfs:gridfs: updating npm dependencies -- mongodb, gridfs-stream...
WARN: Output exceeds 32000 characters
WARN: Output exceeds 32000 characters

[-] Running npm install in the server bundle...


> fibers@1.0.15 install /opt/reaction/dist/bundle/programs/server/node_modules/fibers
> node build.js || nodejs build.js

`linux-x64-46` exists; testing
Binary is fine; exiting

> meteor-dev-bundle@0.0.0 install /opt/reaction/dist/bundle/programs/server
> node npm-rebuild.js


> bcrypt@1.0.2 install /opt/reaction/dist/bundle/programs/server/npm/node_modules/bcrypt
> node-pre-gyp install --fallback-to-build

module.js:327
    throw err;
    ^

Error: Cannot find module '../'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/opt/reaction/dist/bundle/programs/server/npm/node_modules/.bin/node-pre-gyp:15:20)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Function.Module.runMain (module.js:441:10)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bcrypt@1.0.2 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the bcrypt@1.0.2 install script 'node-pre-gyp install --fallback-to-build'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the bcrypt package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-pre-gyp install --fallback-to-build
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs bcrypt
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls bcrypt
npm ERR! There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.meteor/packages/meteor-tool/.1.4.4_2.5g3s0q++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/.npm/_logs/2017-05-23T10_38_29_888Z-debug.log
npm WARN meteor-dev-bundle@0.0.0 No description
npm WARN meteor-dev-bundle@0.0.0 No repository field.
npm WARN meteor-dev-bundle@0.0.0 No license field.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! meteor-dev-bundle@0.0.0 install: `node npm-rebuild.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the meteor-dev-bundle@0.0.0 install script 'node npm-rebuild.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the meteor-dev-bundle package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node npm-rebuild.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs meteor-dev-bundle
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls meteor-dev-bundle
npm ERR! There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.meteor/packages/meteor-tool/.1.4.4_2.5g3s0q++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/.npm/_logs/2017-05-23T10_38_29_936Z-debug.log
The command '/bin/sh -c cd $APP_SOURCE_DIR &&             bash $BUILD_SCRIPTS_DIR/build-meteor.sh &&             bash $BUILD_SCRIPTS_DIR/post-build-cleanup.sh' returned a non-zero code: 1
```

**Comment by prakspcd** [comment]
@jshimko nvm my report, i faced the issue after upgrading to sierra but for a different framework. I posted here by mistake.

If it helps I fixed the build by running 
`sudo launchctl unload /System/Library/LaunchDaemons/org.ntp.ntpd.plist`

**Comment by jshimko** [comment]
@gsc-dev that's the same issue I've been seeing for a while too.  Since it's not the same error as the one this issue was opened for, I've opened a new issue for it (#2333) and included my current understanding of the problem.

**Comment by aaronjudd** [comment]
Closed in favor of #2333 which is the current error I'm experiencing, although the overal "can't build on macOS" remains true...
